shim.go

```diff
@@ -0,0 +1,148 @@
+// Copyright 2017 HyperHQ Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+
+	"github.com/moby/moby/pkg/term"
+	context "golang.org/x/net/context"
+
+	pb "github.com/kata-containers/agent/protocols/grpc"
+)
+
+const sigChanSize = 2048
+
+var sigIgnored = map[syscall.Signal]bool{
+	syscall.SIGCHLD:  true,
+	syscall.SIGPIPE:  true,
+	syscall.SIGWINCH: true,
+	syscall.SIGBUS:   true,
+	syscall.SIGSEGV:  true,
+	syscall.SIGABRT:  true,
+}
+
+type shim struct {
+	containerId string
+	pid         uint32
+
+	ctx   context.Context
+	agent *shimAgent
+}
+
+func newShim(addr, containerId string, pid uint32) (*shim, error) {
+	if agent, err := newShimAgent(addr); err != nil {
+		return nil, err
+	} else {
+		return &shim{containerId: containerId,
+			pid:   pid,
+			ctx:   context.Background(),
+			agent: agent}, nil
+	}
+}
+
+func (s *shim) proxyStdio(wg *sync.WaitGroup) {
+	// don't wait the copying of the stdin, because `io.Copy(inPipe, os.Stdin)`
+	// can't terminate when no input. todo: find a better way.
+	wg.Add(2)
+	inPipe, outPipe, errPipe := shimStdioPipe(s.ctx, s.agent, s.containerId, s.pid)
+	go func() {
+		_, err1 := io.Copy(inPipe, os.Stdin)
+		_, err2 := s.agent.CloseStdin(s.ctx, &pb.CloseStdinRequest{
+			ContainerId: s.containerId,
+			PID:         s.pid})
+		shimLog.Infof("copy stdin err1 %s err2 %s", err1, err2)
```

Comment:
Do we really want to log these `errors` every time? It would be more conventional to do something like:

```
if err1 != nil {
    shimLog.WithError(err1).Warn("failed to copy stdin to pipe")
```

(and the same sort of thing for `err2`).

Same comment applies to the log calls below.